File: windows/ShareWindows/Models/Device.cs
namespace MeshDrop.Models;

public enum DeviceOS
{
    Ios, Android, Macos, Windows, Linux
}

public static class DeviceOSExtensions
{
    public static string ToRaw(this DeviceOS os) => os switch
    {
        DeviceOS.Ios => "ios",
        DeviceOS.Android => "android",
        DeviceOS.Macos => "macos",
        DeviceOS.Windows => "windows",
        DeviceOS.Linux => "linux",
        _ => "linux",
    };

    public static DeviceOS? Parse(string s) => s switch
    {
        "ios" => DeviceOS.Ios,
        "android" => DeviceOS.Android,
        "macos" => DeviceOS.Macos,
        "windows" => DeviceOS.Windows,
        "linux" => DeviceOS.Linux,
        _ => null,
    };
}

/// <summary>
/// 协议规范定义的设备元数据。等价于 Swift `Device`、Kotlin `Device`。
/// </summary>
public sealed record Device(
    string Id,            // 32 hex
    string Name,          // 已 base64url 解码
    DeviceOS Os,
    string? Model,
    string Fingerprint,   // 32 hex
    ushort Port,
    byte ProtocolVersion = 1
)
{
    /// <summary>4 字符分组、空格分隔、大写。</summary>
    public string HumanFingerprint
    {
        get
        {
            var upper = Fingerprint.ToUpperInvariant();
            var groups = new System.Collections.Generic.List<string>();
            for (var i = 0; i < upper.Length; i += 4)
                groups.Add(upper.Substring(i, System.Math.Min(4, upper.Length - i)));
            return string.Join(' ', groups);
        }
    }
}

/// <summary>给 XAML DataTemplate 用的 ViewModel 投影。</summary>
public sealed record DeviceVM(string Id, string Name, string Subtitle, string IconGlyph)
{
    public static DeviceVM From(Device d) => new(
        Id: d.Id,
        Name: d.Name,
        Subtitle: d.Model is null ? d.Os.ToRaw() : $"{d.Os.ToRaw()} · {d.Model}",
        IconGlyph: d.Os switch
        {
            DeviceOS.Ios => "",        // PhoneBook
            DeviceOS.Android => "",
            DeviceOS.Macos => "",      // Tablet/laptop
            DeviceOS.Windows => "",    // Devices
            DeviceOS.Linux => "",
            _ => "",
        });
}

File: apple/Sources/MeshDropKit/Device.swift
import Foundation

public enum DeviceOS: String, Codable, Sendable, CaseIterable {
    case ios, android, macos, windows, linux

    public static var current: DeviceOS {
        #if os(macOS)
        return .macos
        #elseif os(iOS) || os(tvOS) || os(visionOS) || os(watchOS)
        // 协议 v1 的 os 枚举只有 ios/android/macos/windows/linux；
        // tvOS/visionOS/watchOS 暂时统一以 .ios 广告（model 字段携带真实机型）。
        return .ios
        #else
        return .linux
        #endif
    }
}

public struct Device: Identifiable, Hashable, Sendable {
    public let id: String              // 32 hex
    public var name: String            // 已 base64url 解码后的 UTF-8
    public var os: DeviceOS
    public var model: String?
    public var fingerprint: String     // 32 hex
    public var port: UInt16
    public var protocolVersion: UInt8

    public init(
        id: String,
        name: String,
        os: DeviceOS,
        model: String? = nil,
        fingerprint: String,
        port: UInt16,
        protocolVersion: UInt8 = 1
    ) {
        self.id = id
        self.name = name
        self.os = os
        self.model = model
        self.fingerprint = fingerprint
        self.port = port
        self.protocolVersion = protocolVersion
    }

    /// 人眼对齐用：把 32 hex 指纹切成 8 组 4 位、空格分隔、全大写。
    public var humanFingerprint: String {
        let upper = fingerprint.uppercased()
        var out: [String] = []
        var idx = upper.startIndex
        while idx < upper.endIndex {
            let end = upper.index(idx, offsetBy: 4, limitedBy: upper.endIndex) ?? upper.endIndex
            out.append(String(upper[idx..<end]))
            idx = end
        }
        return out.joined(separator: " ")
    }
}
